src/swap/alice.rs

```diff
@@ -123,16 +135,103 @@ pub mod wallet_actor {
     }
 
     #[async_trait::async_trait]
-    impl hbit::Fund for WalletAlice<bitcoin::Wallet, ethereum::Wallet, hbit::PrivateDetailsFunder> {
+    impl<BW, DB> hbit::Fund for WalletAlice<bitcoin::Wallet, BW, DB, hbit::PrivateDetailsFunder>
+    where
+        BW: LatestBlock<Block = ethereum::Block>,
+        DB: db::Load<hbit::CorrectlyFunded> + db::Save<hbit::CorrectlyFunded>,
+    {
+        async fn fund(
+            &self,
+            params: &hbit::Params,
+            beta_expiry: Timestamp,
+        ) -> anyhow::Result<Next<hbit::CorrectlyFunded>> {
+            if let Some(fund_event) = self.db.load(self.swap_id).await? {
+                return Ok(Next::Continue(fund_event));
+            }
+
+            let beta_ledger_time = ethereum_latest_time(&self.beta_wallet).await?;
+            if beta_expiry <= beta_ledger_time {
+                return Ok(Next::Abort);
+            }
+
+            let fund_event = self.fund(&params).await?;
+            self.db.save(fund_event, self.swap_id).await?;
+
+            Ok(Next::Continue(fund_event))
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl<AW, DB, E> herc20::RedeemAsAlice for WalletAlice<AW, ethereum::Wallet, DB, E>
+    where
+        AW: Send + Sync,
+        DB: db::Load<herc20::Redeemed> + db::Save<herc20::Redeemed>,
+        E: Send + Sync,
+    {
+        async fn redeem(
+            &self,
+            params: herc20::Params,
+            deploy_event: herc20::Deployed,
+            beta_expiry: Timestamp,
+        ) -> anyhow::Result<Next<herc20::Redeemed>> {
+            {
+                if let Some(redeem_event) = self.db.load(self.swap_id).await? {
+                    return Ok(Next::Continue(redeem_event));
+                }
+
+                let beta_ledger_time = ethereum_latest_time(&self.beta_wallet).await?;
+                if beta_expiry <= beta_ledger_time {
+                    return Ok(Next::Abort);
+                }
+
+                let redeem_event = self.redeem(&params, deploy_event).await?;
+                self.db.save(redeem_event.clone(), self.swap_id).await?;
+
+                Ok(Next::Continue(redeem_event))
+            }
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl<BW, DB> hbit::Refund for WalletAlice<bitcoin::Wallet, BW, DB, hbit::PrivateDetailsFunder>
+    where
+        BW: Send + Sync,
+        DB: Send + Sync,
+    {
+        async fn refund(
+            &self,
+            params: &hbit::Params,
+            fund_event: hbit::CorrectlyFunded,
+        ) -> anyhow::Result<hbit::Refunded> {
+            loop {
+                let bitcoin_time =
+                    comit::bitcoin::median_time_past(self.alpha_wallet.connector.as_ref()).await?;
+
+                if bitcoin_time >= params.expiry {
+                    break;
+                }
+
+                tokio::time::delay_for(Duration::from_secs(1)).await;
+            }
+
+            let refund_event = self.refund(params, fund_event).await?;
+
```

Comment:
❓ Is there a reason refund is not saved in the DB?

Wouldn't it be important to know when reasoning if we have to do it again on a swap?

Reply:
You're right. I have a TODO in PR #35 to deal with this :)